### PR TITLE
Use test provider for utility tests

### PR DESCRIPTION
**Pull request by alisey**
* Check error codes in the tests
* Use lower-case directory names
* Use `.to.not.throw()` for functions that shouldn't throw
* Throw `ENOTDIR` instead of `ENOENT` in memory fs when trying to create a directory inside of a file. This matches the error code thrown by node fs in the same scenario.
* Remove test that tries to create a directory outside of `/`, because I want to limit all changes to the temp directory.

While utility tests pass on node fs, they cause unrelated tests to fail. For the time being I'm running them on memory fs only.

### CODE REVIEW

File: packages/test-kit/src/sync-utils-contract.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      });
+
+      it('handles special paths gracefully', () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.chdir(tempDirectoryPath);
+        expect(() => fs.ensureDirectorySync('.')).to.not.throw();
+        expect(() => fs.ensureDirectorySync('..')).to.not.throw();
+
+        // This doesn't match node fs, which considers '' to be different from '.',
```
Comment (AviVahl) [inline packages/test-kit/src/sync-utils-contract.ts:57]: memory fs should follow suite rather than being tested separately.
it's a matter of changing `resolvePath` (return '' back as '') and getNode (return null for '')

File: packages/test-kit/src/sync-utils-contract.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        fs.populateDirectorySync(tempDirectoryPath, { 'vesper-bat.txt': '🦇' });
+        const newDirPath = fs.join(tempDirectoryPath, 'vesper-bat.txt', 'habitat');
+
+        expect(() => fs.ensureDirectorySync(newDirPath)).to.throw('ENOTDIR');
+      });
+
+      it('handles special paths gracefully', () => {
+        const { fs, tempDirectoryPath } = testInput;
+
+        fs.chdir(tempDirectoryPath);
```
Comment (AviVahl) [inline packages/test-kit/src/sync-utils-contract.ts:53]: that's why other tests failed. test changed the cwd for the entire process. After this happened, every `nodeFs.resolve` (=== native `path.resolve`) calls used this `tempDirectory` as the `cwd`.

if you delete the folder that process.cwd() points to, both `process.cwd()` and `path.resolve(...)` will throw when called.

the tempDirectoryPath is deleted after the contract finishes, hence the exception.

long story short: is there a way to test this without having to call chdir? or do we want to revert the chdir after the test?

maybe:
```
try {
// test code
} finally{
// restore chdir
}
```
without a catch, any exception will propagate to the test runner, as it should.

File: packages/test-kit/src/sync-utils-contract.ts (hunk 1 above)
Comment (alisey) [reply packages/test-kit/src/sync-utils-contract.ts:57]: I'm pretty sure it shouldn't be fixed at the level of `resolvePath`:

File: packages/test-kit/src/sync-utils-contract.ts (hunk 2 above)
Comment (alisey) [reply packages/test-kit/src/sync-utils-contract.ts:53]: OMG, you're probably right. What do you think about restoring cwd after each test?

File: packages/test-kit/src/sync-utils-contract.ts (hunk 1 above)
Comment (alisey) [reply packages/test-kit/src/sync-utils-contract.ts]: I removed this `ensureDirectorySync('')` assertion for now, as I'm not yet sure how many changes throughout it will require, and don't want to hold up this PR.

Just to be clear, `ensureDirectorySync('')` doesn't match Node's `fs.mkdirSync('')` behavior in both memory fs and node fs.

File: packages/test-kit/src/sync-utils-contract.ts (hunk 2 above)
Comment (alisey) [reply packages/test-kit/src/sync-utils-contract.ts]: Yep, you were right.

**Comment by alisey** [comment]
Crap, Windows is using different error codes.

File: packages/memory/src/memory-fs.ts [hunk 3]
```diff
@@ -225,8 +225,10 @@ export function createBaseMemoryFsSync(): IBaseMemFileSystemSync {
     const parentPath = posixPath.dirname(resolvedPath);
     const parentNode = getNode(parentPath);
 
-    if (!parentNode || parentNode.type !== 'dir') {
+    if (!parentNode) {
       throw createFsError(`${resolvedPath} ${FsErrorCodes.CONTAINING_NOT_EXISTS}`, 'ENOENT');
+    } else if (parentNode.type !== 'dir') {
+      throw createFsError(`${resolvedPath} ${FsErrorCodes.PATH_IS_FILE}`, 'ENOTDIR');
```
[the remark this answers is not in the document]
Comment (alisey) [reply packages/memory/src/memory-fs.ts:231]: Added.

File: packages/utils/src/create-extended-api.ts [hunk 4]
```diff
@@ -86,6 +86,13 @@ export function createExtendedSyncActions(baseFs: IBaseFileSystemSync): IFileSys
       if (parentPath === directoryPath) {
         throw e;
       }
+
+      // Windows also throws ENOENT when trying to create a directory inside of a file,
+      // unlike Mac/Linux that throw ENOTDIR.
+      if (fileExistsSync(parentPath)) {
```
Comment (AviVahl) [inline packages/utils/src/create-extended-api.ts:92]: why is this extra check needed? the recursive call to `ensureDirectorySync(parentPath);` should be the one throwing that exception when it tries to `mkdirSync` that directory.

also, it's important than any change to the sync variants is also applied to the async variants (also in `create-extended-api.ts`)

Comment (alisey) [reply packages/utils/src/create-extended-api.ts:92]: Scenario: creating a directory inside of a file.

After `mkdirSync('/foo.txt/dir')` Mac and Linux throw ENOTDIR, and we're done. But Windows throws ENOENT (same as when the parent directory doesn't exist). Without checking `fileExistsSync(parentPath)` we would recurse and try to create the parent directory `/foo.txt`, and would get `EEXIST`, which is quite misleading, as if `/foo.txt/dir` already exists.

Comment (AviVahl) [reply packages/utils/src/create-extended-api.ts:92]: I've merged #104 so that paths are set on the Error objects themselves.
In the case of `ensureDirectory`, when it re-throws the recursive call's exception, that error will contain the actual path that failed.

Given that this function will soon be a minor wrapper to `mkdir(p, {recursive: true})`, I wouldn't worry too much about the error codes for the edge cases. Instead I rather it be optimized for the common case, and later let the error codes from node's native recursive implementation to propagate. see #103 for script + output on Linux/Windows.
